All deprecations on AtomGroup and friends getters and setters.

dotsdl opened the pull request:
The new topology system for #363 will arrive in 0.16.0, and this removes
redundant getters and setters of various `Atom`, `Residue`, and
`Segment` properties. These methods predated the properties that replace
them, and this commit features deprecation warnings for each with
instructions on what to use instead.

Some `Atom` properties, such as `number`, `pos`, `serials`, have themselves been
replaced in the new scheme (`index`, `position`, `id`, respectively).
The centroid method has also been removed.

<strike>The methods `AtomGroup.translate` and `AtomGroup.rotateby` no longer
take tuples of `AtomGroup` objects as optional arguments, but only take
explicit vectors. Because `Residue`, `ResidueGroup`, `Segment`, and
`SegmentGroup` are no longer `AtomGroup`s themselves, leaving this in
creates a minor convenience at the cost of method complexity. The
documentation is correspondingly complex, and in the course of updating
the docstrings to numpy-style for #363, simplifying the method also
simplified the documenation.</strike>

Also deprecated `as_Universe` function. This has yet to be discussed on
the `issue-363` branch, but discussion should happen on what its purpose
is.

Also addressed in this commit:

- [x] In #363, `__getattr__` behavior for the Groups
have changed to a slightly different scheme. No warnings have been added
for this change to the `__getattr__` methods yet. The new scheme goes
as::

    SegmentGroups __getattr__ can select by segment name
    Segment __getattr__ can select by residue name
    ResidueGroup __getattr__ can select by residue name
    Residue __getattr__ can select by atom name
    AtomGroup __getattr__ can select by atom name

Should only require the following in `AtomGroup.__getattr__`, or
pehaps override in `ResidueGroup` and add after a `super`:

```python
if isinstance(self, ResidueGroup):
    warnings.warn("In version 0.15.0 this will select "
            "residue names, not atom names ", DeprecationWarning)
```

- [x] Because `Residue`, `ResidueGroup`, `Segment`, and
`SegmentGroup` are no longer `AtomGroup`s themselves, there are methods
such as `Segment.names` that now longer exist (one would use
`Segment.atoms.names` instead). Something like:

```python
if isinstance(self, Segment):
    warnings.warn("In version 0.15.0, use `segment.atoms.names "
                  "instead.", DeprecationWarning)
```

will be needed. And since calling `SegmentGroup.names` no longer gives a single ``numpy`` array of all atom names in each segment shoved together, but instead a list of arrays, one for each segment, this will need to be noted. A decorator will be best.


- [x] I left bonds things alone as it's not my area of expertise. Any deprecations on that front should be placed by @richardjgowers.

orbeckst commented:
Quick comment: I would want to keep the "slot in groups of atoms for `translate` and `rotate`" because I  don't want to needlessly break old functionality.
Related note: all the new XGroups should still mostly *behave* like old `AtomGroup` (e.g. `centroid()`).

dotsdl commented:
@orbeckst I will reinstate the conveniences built into these transformation methods.

richardjgowers commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -3336,6 +3378,7 @@ def __init__(self, name, id, atoms, resnum=None):
         ##if not Residue._cache.has_key(name):
         ##    Residue._cache[name] = dict([(a.name, i) for i, a in enumerate(self._atoms)])
 
+    @deprecate(message=_FIFTEEN_DEPRECATION)
```
Can these not get moved across?

dotsdl replied:
I think these things are really fragile. They depend on particular atom names (however customary), and the implementation also seems a bit frail, depending on residue id being in order for the chain. I'd rather mark them as deprecated and then reverse the decision later if we think they can be included with confidence.

They've been included for now among the ``Atomnames`` TopologyAttr in the 363 branch (dbe6d061e35f45104317e2c3662601f19b9ea44e), but they don't work at the moment since they depend on Residue ``__getitem__`` working with atom names in a funny way. This isn't implemented yet, and I'm not sure it should be, but it is possible. See [``__getattr__`` for Residues](https://github.com/MDAnalysis/mdanalysis/blob/dbe6d061e35f45104317e2c3662601f19b9ea44e/package/MDAnalysis/core/topologyattrs.py#L278) for an example of how.

orbeckst replied:
"Fragile" is a very relative term. They work for pretty much anyone who cares about them, i.e. if you have protein simulations then there's a very good chance that this works because the naming is de-facto standard for the backbone atoms (termini is more varied but not as varied between the force fields and PDB that it wouldn't make sense to catch them). It's easy to take the developer's high-road and only provide functionality that "never ever will break" (haha) but that misses the point that most users are quite happy to get the work done in 90% of cases and would simply appreciate a good error message in the remaining 10%.

orbeckst replied:
... so, I don't want to deprecate them.

kain88-de replied:
Yeah rather leave a note in the doc string so people can look into it if it fails.

dotsdl commented:
Retargeted to [0.14.1](https://github.com/MDAnalysis/mdanalysis/milestones/0.14.1%20-%20Bugfixes). Between now and BPS I have zero time to finish this up, and I think we need a bit more progress on [issue-363](https://github.com/MDAnalysis/mdanalysis/tree/issue-363) given discussion in #599 to make it clear exactly what is changing as much in advance as possible.

khuston commented:
I found a potential use for `as_Universe`, though it's not very elegant. If the user wants to manually add a topology attribute, they can

    u.add_TopologyAttr(mda.core.topologyattrs.Masses(np.ones(len(u.atoms))))

And then `u.atoms.masses` works, but `u.atoms[0].mass` results in `AttributeError` because the masses are not propagated from TopologyGroup to TopologyObjects. This is fixed if instead the user does

    u.add_TopologyAttr(mda.core.topologyattrs.Masses(np.ones(len(u.atoms))))
    u = mda.core.universe.as_Universe(u)

Then both `u.atoms.masses` and `u.atoms[0].mass` work. This feels a bit hackish, and perhaps something like `as_Universe` should be performed after the user-accessible `add_TopologyAttr` is called.

Correct me if there is a better way of manually adding an attribute to the topology that gets propagated down to the individual atoms.

richardjgowers commented:
@khuston hmm, adding the attribute should propogate it without your hack... I'll have a poke around

khuston commented:
I'm sorry -- I'm mistaken, and the `as_Universe` is unnecessary. I must have been using the wrong version when I tried that.

kain88-de commented:
@dotsdl what is still needed to get the depcreations merged?

dotsdl commented:
@kain88-de I'll be addressing the comments made here and elsewhere (#599) and have this PR ready for merge by the end of the week. I've got this whole week for development, and this is among the things at the top of my list.

dotsdl commented:
Added more deprecations; see top comment for updates. Ready for review again and hopeful for merge.

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -457,8 +457,84 @@
 _SINGULAR_PROPERTIES = {v: k for k, v in _PLURAL_PROPERTIES.items()}
 
-_FIFTEEN_DEPRECATION = "This will be removed in version 0.15.0"
+_FIFTEEN_DEPRECATION = "This will be removed in version 0.16.0"
```
Can you run a global replace to `_SIXTEEN_DEPRECATION`? In case you are using emacs and projectile the short cut it `C-p r`

kain88-de replied:
Or just in this file since all changes are in one file.

dotsdl replied:
Yeah, I'll make the change.

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -671,6 +749,7 @@ def force(self, vals):
             raise NoDataError("Timestep does not contain forces")
 
+    @deprecate(message="{}; use `position` property instead".format(_FIFTEEN_DEPRECATION))
     def centroid(self):
```
what? the `centroid` isn't the same as the `position` array in my understanding.

dotsdl replied:
This is for an ``Atom``, for which ``centroid`` is just its position. Unless there's a reason to keep this, I think it's unnecessary at best, confusing at worst. Happy to hear dissenting opinions, though.

kain88-de replied:
I didn't see it was for an `Atom` then yes it should be removed.

richardjgowers replied:
I'm not in favour of centroid per se, but I'd rather keep things focused rather than throwing in changes we'd like.

richardjgowers replied:
I mean keep the PR focused, `centroid` deserves a fair trial :D

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -685,6 +764,7 @@ def universe(self):
             return self._universe
 
     @universe.setter
+    @deprecate(message="{}; Atoms will not be able to leave their Universes.".format(_FIFTEEN_DEPRECATION))
```
out of curiosity how is this triggered?

dotsdl replied:
When a user tries to set the universe an ``Atom`` is a member of, this warning will appear. If you're asking where this is used in the codebase, no idea. In the new scheme though it doesn't make sense to do this.

richardjgowers replied:
It used to be that Parser created Atoms, then the Universe assigned itself to each Atom afterwards in init

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -3660,14 +3803,26 @@ def __init__(self, name, residues):
         self._cls = ResidueGroup
 
     @property
+    @deprecate(message="{}; use `segid` property instead".format(_FIFTEEN_DEPRECATION))
```
This conflicts with the doc string and `segid` doesn't work currently

```python
# works
u.atoms.segments[0].id
# doesn't work
u.atoms.segments[0].segid
```

kain88-de replied:
This is important to me. When we issue deprecation messages then the suggested alternative should work.

dotsdl replied:
Good catch. Added a ``segid`` property to ``Segment`` to shore this up.

kain88-de replied:
ideally this should be done with all deprecations.

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -3660,14 +3804,35 @@ def __init__(self, name, residues):
+        self.name = x
+
+    @property
+    @deprecate(message="{}; use `segid` property instead".format(_SIXTEEN_DEPRECATION))
```
Is git messing up or are you really *adding* deprecated features here?

dotsdl replied:
I think git's messing up. I only added the ``segid`` property, which for the moment refers to segment ``name`` internally.

dotsdl replied:
Since ``id`` also uses name (see left of screen), it would be really bad if ``name`` wasn't already there.

richardjgowers commented:
So bonds won't work any differently in the future, apart from the fact they're now only accessible from Atom/AtomGroup.  I've added deprecations for this

dotsdl commented:
Okay, so I hope nobody uses ``Atom.id`` as they currently exist because they apparently get assigned by residues. There is no corresponding mechanism in the new system (there shouldn't be; it's very unnecessary and just asking for staleness issues). 

I'm shoring up the suggestion to use ``AtomGroup.ids`` instead of ``AtomGroup.serials`` in the deprecation warning by giving the serials back if they exist, but otherwise giving back ``AtomGroup.indices``. I'm also removing the setting of ``Atom.id`` [by residues as they are built](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/AtomGroup.py#L3329) because we'd like this to work the same for individual ``Atom``s as it does for ``AtomGroup``.

dotsdl commented:
Alright, we should be set, then. Ready to merge if there is nothing further.

orbeckst commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -4671,12 +4865,14 @@ def trajectory(self, value):
         del self._trajectory  # guarantees that files are closed (?)
         self._trajectory = value
 
+    @deprecate(message=_SIXTEEN_DEPRECATION)
```
Will we not need anchoring anymore?

dotsdl replied:
What does it do right now? It doesn't exist as a concept in 363, so I removed it, but is it necessary given that AtomGroups no longer point to Atoms, and that Atoms no longer exist without a Universe?

orbeckst replied:
@mnmelo implemented it; it's required for pickling AtomGroup IIRC.

On 30 Mar, 2016, at 16:04, David Dotson wrote:

> What does it do right now? It doesn't exist as a concept in 363, so I removed it, but is it necessary given that AtomGroups no longer point to Atoms, and that Atoms no longer exist without a Universe?

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

orbeckst commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -4715,6 +4915,7 @@ def _matches_unpickling(self, anchor_name, n_atoms, fname, trajname):
             return False
 
 
+@deprecate(message=_SIXTEEN_DEPRECATION)
```
This *might* get un-deprecated... as you said in your summary.

dotsdl replied:
Yeah, it doesn't make sense at the moment in 363, so best to deprecate and then not remove it at all if it ends up being needed.

orbeckst commented:
Good work @dotsdl , as far as I can tell. But I am not deeply involved in #363 so perhaps could @kain88-de or @richardjgowers take charge of the final merge? 

Should we already have a discussion with @tylerjereddy regarding new fixers for "ten2sixteen"?

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -457,7 +457,84 @@
 _SINGULAR_PROPERTIES = {v: k for k, v in _PLURAL_PROPERTIES.items()}
 
-_FIFTEEN_DEPRECATION = "This will be removed in version 0.15.0"
+_SIXTEEN_DEPRECATION = "This will be removed in version 0.16.0"
```
Looks like the deprecation warning was something older. I thought you introduced it in this PR. Did we by accident reschedule some deprecations now? Looking over the diff yes we did reschedule some deprecations.

Could you go over that again @dotsdl ?

We need a little bit more formal way to deal with deprecations and how to check that we remove functions for the targeted releases.

orbeckst replied:
On 31 Mar, 2016, at 00:17, kain88-de wrote:

> We need a little bit more formal way to deal with deprecations and how to check that we remove functions for the targeted releases.

Do you mean like adding a test that *fails* if a function still exists in a release?

In this particular case I have rather been advocating tests that test what we want to keep https://github.com/MDAnalysis/mdanalysis/issues/703

dotsdl replied:
The reason this went from "FIFTEEN" to "SIXTEEN" was because we were originally planning for the topology refactor to land in 0.15.0, with the release we're working toward now being a point release (0.14.1). Since we decided to release next as 0.15.0, this moved the topology refactor merge to 0.16.0.

Does that clarify things? I realize that it would be nice to have a more formal process for managing deprecations, but I think a major internal refactor like this is a rarity and this is about as clean as it probably gets.

kain88-de replied:
If this change for the other deprecations is intentional I'm fine with it. 

What I was referring that we currently clean up *a lot* and deprecate old stuff all over. So much that I forget what is supposed to be deprecated when myself. Just a list that we try to keep up to date for us devs to see what deprecations to check for when a release is happening.

orbeckst replied:
On 31 Mar, 2016, at 11:04, David Dotson wrote:

> The reason this went from "FIFTEEN" to "SIXTEEN" was because we were originally planning for the topology refactor to land in 0.15.0, with the release we're working toward now being a point release (0.14.1).

Er, I think the next release WILL be 0.15.0... we already have too much new stuff (see CHANGELOG). It's just the current version tag is still 0.14.1.

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

dotsdl replied:
@orbeckst that's what I said. The changes were too substantial to warrant a point release. So, there's nothing to worry about with why these all went from being deprecations of things that will go missing in 0.15.0 to things that will go missing in 0.16.0. We just shifted the big #363 merge a release.

kain88-de commented:
Looks good to me. I've you add it to the CHANGELOG I think we are good to go.

Since all deprecations already have a replacement I think @tylerjereddy could make changes for a `ten2sixteen` script that people can already use to be ahead of the curve.

dotsdl commented:
Hold up...I missed some things. For one thing, `Residue` doesn't have `resid` or `resname`, but instead using `id` and `name`. *sigh*

Adding properties for these things, and the appropriate deprecation for the alternative that should be used.

dotsdl commented:
So, did a lot in the last commit (b399e183430b564af3c5da35400075ce2667556e). See log for details.

What needs doing: we get deprecation warnings on load of a `Universe` with a PSF + DCD. This is probably because the readers for these might be calling deprecated components of `MDAnalysis.core.AtomGroup.py`. Could someone that's not me try loading up a bunch of different Universes and changing the problem spots to using our "recommended" accessors/getters?

tldr: our own code shouldn't throw deprecation warnings. Only user code should.

richardjgowers commented:
Ok I'm reviewing this

tylerjereddy commented:
So, we want a separate `ten2sixteen` or we want to update the old one to have this new name instead of `ten2eleven`? Are the changes all summarized nicely in a single place? I guess a lot of them are mentioned or can be deduced from this PR, though end users would likely want a summary somewhere anyway.

richardjgowers commented:
@dotsdl loading PSF/DCD universe doesn't throw warnings now, except.....

The anchoring stuff.  I read through it all, and there's no reason why it won't work in the new scheme.  I'm not sure how useful it is to pickle AtomGroups? but it should still work as best I can tell.  Maybe remove these deprecations and raise an Issue if we want to axe this feature?

dotsdl commented:
@richardjgowers I don't think it makes any sense to pickle AtomGroups in the new scheme, since they are useless without a ``Topology``, and therefore really a ``Universe``. I think the idea of pickling AtomGroups should be abandoned in favor of making ``Universe`` pickle just fine so it can be used directly as arguments for functions using multiprocessing. Of course this means that we must be clear what pickling a ``Universe`` means (probably fully serialized Topology (#643) plus paths to trajectory(s)).

But that's another discussion. Let's de-deprecate the anchoring machinery so it stops throwing warnings from internal use. I doubt users are using this stuff directly, though I could be surprised. 

We should put a deprecation warning on AtomGroup's  `__getstate__` and `__setstate__` methods if we want to discourage pickling them.

richardjgowers commented:
@tylerjereddy we started making a list [here](https://github.com/MDAnalysis/mdanalysis/wiki/Issue363-Changes) but it needs expanding a lot.  I'll try and fill it out today

dotsdl commented:
Okay, after some last fixes on some omissions, I think we're finally ready. CHANGELOG updated, too. These deprecation warnings should help us out when adapting the existing tests to the new API conventions, since currently the test suite throws them everywhere.

orbeckst commented:
On 2 Apr, 2016, at 09:46, Tyler Reddy wrote:

> So, we want a separate ten2sixteen or we want to update the old one to have this new name instead of ten2eleven?

Assuming that there's no harm in running "ten2sixteen" on >0.11 code,  I'd say make a new ten2sixteen and remove ten2eleven ... we just mustn't forget to update wiki etc.

dotsdl commented:
Anything else to address here? Otherwise can we merge?

kain88-de commented:
This results in a lot of warning messages now about deprecations.  It should be easy enough to grep for the words and replace them but I lost track what they all were. @dotsdl  can you make a list of the deprecated methods and their preferred alternative, then I can change them across the whole codebase (emacs has a function for that :wink:)

kain88-de commented:
Plus I get deprecation warning when I use the preferred methods `ag.positions` triggers a warning that `get_positions` is deprecated.

richardjgowers commented:
@kain88-de I'm working on this a little.  positions is because it goes through get_positions, so it needs to get redone so that get_positions goes via positions.  Ideally we should be able to run the tests with no deprecation warnings

kain88-de commented on package/MDAnalysis/core/AtomGroup.py:
```diff
@@ -1143,6 +1234,7 @@ def n_segments(self):
         return len(self.segments)
 
     @property
+    @warn_atom_property
```
which one should be used instead in the future?

dotsdl replied:
``*Group.atoms.indices``

Indices always refer to atom indices (zero-based, and always unique among atoms), no matter what level called from. In order to get back indices as they currently work from all levels (except AtomGroup) in the new system, one must prefix with ``.atoms``.
